### PR TITLE
feat: new wait for idle

### PR DESCRIPTION
**Pull request by dimaqq**
Pull the new Model.wait_for_idle implementation from #1104 

- [x] pull in the unit tests
- [x] pull in the implementation
- [x] new unit tests pass
- [x] all unit tests pass
- [x] feature flag
- [x] run integration tests with and without the feature flag
- [x] implement idle/ready distinction properly
- [x] merge #1242 and rebase

### CODE REVIEW

**Comment by dimaqq** [comment]
@james-garner-canonical I'd appreciate an early review.

I'm expecting integration tests to pass 🎉 

There's still some refactoring to be done... and I'm happy to hear your thoughts about that and areas to improve in general.

**Comment by dimaqq** [comment]
Transient error:
`juju.errors.JujuUnitError: 'docker-registry/0' workload has errored: 'hook failed: "install"'`

I’ve restarted that one job

File: juju/model.py [hunk 1]
```diff
@@ -3212,6 +3245,222 @@ def _raise_for_status(entities: dict[str, list[str]], status: Any):
...
+                "wait_for_active is deprecated; use status",
+                DeprecationWarning,
+                stacklevel=3,
+            )
+            status = "active"
+
+        _wait_for_units = (
+            wait_for_at_least_units if wait_for_at_least_units is not None else 1
+        )
+        apps = apps or list(self.applications)
```
Comment (james-garner-canonical) [inline juju/model.py]: I see there was some type checking here in the old implementation that raised a `JujuError` if `apps` wasn't a `list` of `str` (or `None`), with a note that passing a `str` for the `apps` argument was a common error.

Looks like now if you passed a `str` or any other iterable, `_check_idle` would `return False` on line 3330 instead, leading to either timeout or run forever. Well, unless your  app names were all one letter and you passed a string concatenating their names, that would work like you'd passed them in a list.

Worth adding the `JujuError` back? And maybe a unit test that the new implementation handles bad arguments the same way that the old implementation did?

File: juju/model.py [hunk 2]
```diff
@@ -3212,6 +3245,222 @@ def _raise_for_status(entities: dict[str, list[str]], status: Any):
...
+
+        # Order of errors:
+        #
+        # Machine error (any unit of any app from apps)
+        # Agent error (-"-)
+        # Workload error (-"-)
+        # App error (any app from apps)
+        #
+        # Workload blocked (any unit of any app from apps)
+        # App blocked (any app from apps)
```
Comment (james-garner-canonical) [inline juju/model.py]: I see that this order was previously defined in the `_raise_for_status` inner function.
(I also note the use of Workload instead of Unit, though a `JujuUnitError` is still what's raised).

I think returning and raising early is a much nicer implementation. Are there any concerns with the order of operations not matching exactly? I think where the order of operations is timing related (when exactly in time we query juju), it doesn't matter so much -- the variance in status due to timing washes out the difference there I think. What are your thoughts here?

I like this face btw (-"-)

File: juju/model.py [hunk 3]
```diff
@@ -3212,6 +3245,222 @@ def _raise_for_status(entities: dict[str, list[str]], status: Any):
...
+                    "Waiting for app %r units %s/%s",
+                    app_name,
+                    len(ready_units),
+                    _wait_for_units,
+                )
+                return False
+
+        if busy := [n for n, t in idle_times.items() if expected_idle_since < t]:
+            logger.info("Waiting for %s to be idle enough", busy)
+            return False
```
Comment (james-garner-canonical) [inline juju/model.py]: Suggestions:
- rename `idle_times` to `last_active` or `last_non_idle` (more obvious what we're tracking here)
- could we use a property or something to exfiltrate this state between calls instead of a mutable argument? e.g. `self._last_active_unit_times` (any concurrency concerns with this?)
- I like the one liner, but to understand it I mentally expand to more verbose and explicit names, (and swap conditional direction)
```py
busy_units = [
    unit_name
    for unit_name, last_active_time
    in last_active.items()
    if last_active_time > required_idle_since
]
if busy_units: ...
```

File: juju/model.py [hunk 4]
```diff
@@ -3212,6 +3245,222 @@ def _raise_for_status(entities: dict[str, list[str]], status: Any):
...
+
+        return True
```
Comment (james-garner-canonical) [inline juju/model.py]: Below I've pasted my understanding of the new and original wait for idle logic. Differences in order of operations are:
- original could raise errors before all apps are available, new can't
- original could return early once enough units were ready (assuming no errors had been accumulated in that pass through the apps) -- new will check for all errors before checking if units are ready (an improvement I think)

I suspect that both of these divergences are fine.

---

new

0. do all apps exist? if not, stop and request to check again later
1. check all apps and units for errors in the appropriate order, raising immediately if the user cares
2. for each app, figure out which units are ready -- if there are less than we're waiting for, stop and request to check again later
3. check if units have been idle for long enough -- if not, stop and request to heck again later
4. we're done

original

0. iterate over each app
1. does the app exist yet? if not, stop processing this app and request to iterate again
2. is the app in error/blocked and is that a problem? record it if so
3. are we waiting for enough units, and if so are there enough?
    - if there aren't, stop processing this app
    - if there are, stop processing apps entirely (but we'll start again if any earlier apps requested it -- not any later ones though -- smells like a bug)
4. now check the app's units for errors/blocked and record it if the user cares
5. now figure out if the app status means we need to start again
6. after stopping iteration
    - if any errors were recorded, raise the highest priority one
    - if any apps needed to start again, restart iteration, otherwise we're done

File: tests/unit/test_wait_for_idle.py [hunk 5]
```diff
@@ -0,0 +1,540 @@
+# Copyright 2024 Canonical Ltd.
```
Comment (james-garner-canonical) [inline tests/unit/test_wait_for_idle.py]: I like the new unit tests, just a note about testing the case where `apps` is a `str` since old implementation raises an error here that I don't believe the new implementation does (instead silently timing out?)

File: tests/unit/test_wait_for_idle.py [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+    kwargs["apps"] = ["hexanator"]
+    kwargs["wait_for_at_least_units"] = 2
+    kwargs["_wait_for_units"] = 2
+    kwargs["status"] = "active"
+
+    idle = await model_fake(full_status_response)._check_idle(**kwargs)
+    assert not idle
+
+
+# FIXME eventual timeout
```
Comment (james-garner-canonical) [inline tests/unit/test_wait_for_idle.py]: Not clear to me what this is referring to

File: juju/unit.py [hunk 7]
```diff
@@ -15,7 +15,9 @@
 
 
 class Unit(model.ModelEntity):
-    name: str
+    @property
+    def name(self) -> str:
+        return self.entity_id
```
Comment (james-garner-canonical) [inline juju/unit.py:20]: Was this a type annotation error? name never actually set outside our tests?

File: juju/client/facade.py [hunk 8]
```diff
@@ -479,37 +479,48 @@ def ReturnMapping(cls):  # noqa: N802
...
 
+@overload
+def _convert_response(response: dict[str, Any], *, cls: type[SomeType]) -> SomeType: ...
+
+
+@overload
+def _convert_response(response: dict[str, Any], *, cls: None) -> dict[str, Any]: ...
+
+
+def _convert_response(response: dict[str, Any], *, cls: type[Type] | None) -> Any:
```
Comment (james-garner-canonical) [inline juju/client/facade.py:498]: I see why you split this out, very helpful for testing. I think the overload isn't quite correct according to the implementation, since we can get an error type depending on the response contents, and the code will return a list if the `cls` was defined by a schema with `"type": "array"`.
```py
if TYPE_CHECKING:
    from juju.client._definitions import Error
@overload
def _convert_response(
    response: dict[str, Any], *, cls: type[SomeType]
) -> SomeType | list[SomeType] | Error: ...
@overload
def _convert_response(response: dict[str, Any], *, cls: None) -> dict[str, Any]: ...
def _convert_response(response: dict[str, Any], *, cls: type[Type] | None) -> Any:
```
I think `Any` is fine as the return type of the concrete implementation to avoid touching the implementation itself too much. Could change `dict[str, Any]` to `dict[str, JSONData]` but not a big deal.

**Review by james-garner-canonical — COMMENTED** [review]
Looks great, so much nicer than the previous implementation. Really cool to have the tests passing on both. Was wait_for_idle tested much previously?

File: juju/model.py (hunk 2 above)
Comment (dimaqq) [reply juju/model.py]: The client code is run against a snapshot of Juju state, so order can be deterministic, as far as Juju allows that.

The reason I wanted to preserve the semantics of the exceptions this method raises is that there are existing integration tests `with pytest.raises(JujuUnitError)` that I would rather not accidentally break.

Arguably those tests should catch the exception by superclass, but they are what they are.

File: juju/model.py (hunk 3 above)
Comment (dimaqq) [reply juju/model.py]: Let me think about that.
last active time is not actually tracked.
What is tracked is "unit X was continuously idle since sampled at 11:22:33", for some degree of "continuously" obv.

File: juju/model.py (hunk 4 above)
Comment (dimaqq) [reply juju/model.py]: I'm refactoring this mess... wait for it!

File: juju/unit.py (hunk 7 above)
Comment (dimaqq) [reply juju/unit.py:20]: There are, weirdly two sources of truth for the name:
- any ModelEntity, when created is pinned by entity id.
- then state updates are dispatched to safe data by entity id.
- safe_data["name"] exists and equals entity id because of the dispatch.
- __getattr__(self, "name") used to fetch it from safe_data

I'm basically making this a little more sane.
It's true that I didn't have to make this change in this very PR, but it helped me with types.... or tests for something.

Comment (james-garner-canonical) [reply juju/unit.py:20]: I have no objections to making the change, it just jumped out at me. Thanks for explaining

File: juju/model.py (hunk 3 above)
Comment (james-garner-canonical) [reply juju/model.py]: Ah yep, so `last_active_time` is not really an accurate name. `idle_since` maybe

File: juju/model.py (hunk 2 above)
Comment (james-garner-canonical) [reply juju/model.py]: 100% agree that raising the exact same exceptions is best for backwards compatibility

File: juju/model.py (hunk 1 above)
Comment (dimaqq) [reply juju/model.py]: done

File: juju/model.py (hunk 3 above)
Comment (dimaqq) [reply juju/model.py]: done

File: juju/model.py (hunk 4 above)
Comment (dimaqq) [reply juju/model.py]: Please have a look now :)

File: tests/unit/test_wait_for_idle.py (hunk 5 above)
Comment (dimaqq) [reply tests/unit/test_wait_for_idle.py]: done

File: tests/unit/test_wait_for_idle.py (hunk 6 above)
Comment (dimaqq) [reply tests/unit/test_wait_for_idle.py]: moved and expanded the list of missing tests

File: juju/client/facade.py (hunk 8 above)
Comment (dimaqq) [reply juju/client/facade.py:498]: Let me think about that...
I thought that if there was an error, it would get raised, rather than returned.
But maybe I misread the code.

**Comment by dimaqq** [comment]
@james-garner-canonical the big refactor is done, I hope you'll like the result :)

still a draft, as some unit tests are missing and I have one internal concern, but should be ready for review :)

File: juju/client/facade.py (hunk 8 above)
Comment (james-garner-canonical) [reply juju/client/facade.py:498]: There's something like
```py
if 'error' in reply:
    cls = CLASSES['Error']
... # later
result = cls.from_json(reply['response'])
return result
```
And `Error` is a `Type` subclass in `_definitions.py` -- I don't think it raises on being instantiated or anything unless the `response` is messed up

File: juju/model/idle.py [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+            rv.units.add(unit_name)
+            assert unit.agent_status
+            assert unit.workload_status
+
+            if unit.agent_status.status != "idle":
+                continue
+            if status and unit.workload_status.status != status:
+                continue
+
+            ready_units.append(unit)
```
Comment (james-garner-canonical) [inline juju/model/idle.py]: I guess `ready_units` is a leftover that can be removed?

File: juju/model/__init__.py [hunk 10]
```diff
@@ -3204,14 +3248,87 @@ def _raise_for_status(entities: dict[str, list[str]], status: Any):
...
+            raise ValueError("apps must be a Sequence[str]")
+
+        idle_loop = idle._loop(
+            apps=apps,
+            timeout=timeout,
+            wait_for_exact_units=wait_for_exact_units,
+            wait_for_units=wait_for_units,
+            idle_period=idle_period,
+        )
+        next(idle_loop)  # Prime the generator
```
Comment (james-garner-canonical) [inline juju/model/__init__.py]: Can we eliminate the need for this by just setting `status` to `None` unconditionally on the first line of `_loop` instead of using `= yield`?

(But see comments on `def _loop` -- I think a regular object would be a lot easier to follow internally than a generator)

File: juju/model/idle.py [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+    wait_for_exact_units: int | None = None,
+    wait_for_units: int,
+    idle_period: float,
+) -> Generator[bool, _CheckStatus | None, None]:
+    """The outer, time-dependents logic of a wait_for_idle loop."""
+    status: _CheckStatus | None = yield False  # None at first
+    deadline = None if timeout is None else time.monotonic() + timeout
+    idle_since: dict[str, float] = {}
+
+    while True:
```
Comment (james-garner-canonical) [inline juju/model/idle.py]: I think the purpose of having a generator here is to encapsulate some initially provided parameters and some internal state (`idle_since`), is that right? If so, I think it would be worth making this a regular class instead. Example below of what I think should be pretty much a drop in equivalent which seems easier to understand (to me)
```py
class _loop:  # class IdleLoop
    def __init__(
        self,
        *,
        apps: Sequence[str],
        timeout: float  | None,
        wait_for_exact_units: int | None = None,
        wait_for_units: int,
        idle_period: float,
    ):
        self.apps = apps
        self.deadline = None if timeout is None else time.monotonic() + timeout
        self.wait_for_exact_units = wait_for_exact_units
        self.wait_for_units = wait_for_units
        self.idle_period = idle_period
        self.idle_since: dict[str, float] = {}

    def send(self, status: _CheckStatus | None) -> bool:  # def run/check
        # body of while True loop with self. where appropriate
```

File: juju/model/idle.py [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+        #
+        # However, the "idle_since" check may need to be updated
+        # Right now, a "busy" unit is stamped with ready := since right now
+        #
+        # This is fine for idle_period > 0
+        # But may be problematic for idle_period===0 and wait_for_units==0
+        # (there's a single use case for that in the wild)
+        #
+        # Specifically, a unit in an error state is not ready
+        # But it gets stamped with "now" and seems "idle enough"
```
Comment (james-garner-canonical) [inline juju/model/idle.py]: Interesting. What are the semantics of the original `wait_for_idle` here, and how important is it to match them? The behaviour you've described seems to be clearly buggy. A simple solution might be to stamp a "busy" unit with ready := since some time in the future.

File: juju/model/idle.py [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+                n for n in status.ready_units if n.startswith(f"{app_name}/")
+            ]
+            if len(ready_units) < wait_for_units:
+                logger.info(
+                    "Waiting for app %r units %s >= %s",
+                    app_name,
+                    len(status.ready_units),
+                    wait_for_units,
+                )
+                rv = False
```
Comment (james-garner-canonical) [inline juju/model/idle.py]: Since the loop never sets `rv =True` I guess we could break any time we set `rv = False`? But it's cheap since we're just iterating over data we've already queried from Juju, and the logging is nice, so I'm -1 on my own suggestion here ...

File: juju/model/idle.py [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+                    wait_for_exact_units,
+                )
+                rv = False
+
+        # FIXME possible interaction between "wait_for_units" and "idle_period"
+        # Assume that we've got some units ready and some busy
+        # What are the semantics for returning True?
+        if busy := [n for n, t in idle_since.items() if t > expected_idle_since]:
+            logger.info("Waiting for %s to be idle enough", busy)
+            rv = False
```
Comment (james-garner-canonical) [inline juju/model/idle.py]: My first thought is something like this:
```py
ready = [n for n, t in idle_since.items() if t < expected_idle_since]
if len(ready) < wait_for_units:
    logger.info(...)
    rv = False
```
But maybe it would make more sense to instead use `idle_since` to filter units out of `ready_units` before the `for app in apps` above (line 76) -- or have that loop use `idle_since` internally if you want the logging to be able to distinguish between whether a unit is ready at all vs ready for long enough?

File: juju/model/__init__.py [hunk 15]
```diff
@@ -3064,16 +3103,21 @@ async def wait_for_idle(
             wait_for_at_least_units if wait_for_at_least_units is not None else 1
         )
 
-        timeout = timedelta(seconds=timeout) if timeout is not None else None
-        idle_period = timedelta(seconds=idle_period)
+        timeout_ = timedelta(seconds=timeout) if timeout is not None else None
+        idle_period_ = timedelta(seconds=idle_period)
```
Comment (james-garner-canonical) [inline juju/model/__init__.py:3108]: Names like this always strike me as potentially very error prone for future changes. Is the rename to get the types right?

File: juju/model/__init__.py [hunk 16]
```diff
@@ -3064,16 +3103,21 @@ async def wait_for_idle(
...
+            raise ValueError("apps must be a Sequence[str] and not a plain type")
+
+        apps_ = apps or list(self.applications)
+
+        if not isinstance(apps_, collections.abc.Sequence):
+            raise ValueError("apps must be a Sequence[str]")
+
+        if any(not isinstance(o, str) for o in apps_):
+            raise ValueError("apps must be a Sequence[str]")
+
```
Comment (james-garner-canonical) [inline juju/model/__init__.py]: So this is the first actual change this PR makes to old wait for idle. `ValueError` seems appropriate here (or `TypeError`), but is it ok to switch from raising a `JujuError`? Presumably no one is catching these currently in their tests anyway so it's probably ok

File: juju/model/__init__.py [hunk 17]
```diff
@@ -3064,16 +3103,21 @@ async def wait_for_idle(
...
-        if apps is not None and (
-            not isinstance(apps, list) or any(not isinstance(o, str) for o in apps)
-        ):
-            raise JujuError(f"Expected a List[str] for apps, given {apps}")
 
-        apps = apps or self.applications
+        if isinstance(apps, (str, bytes, bytearray, memoryview)):
+            raise ValueError("apps must be a Sequence[str] and not a plain type")
+
+        apps_ = apps or list(self.applications)
```
Comment (james-garner-canonical) [inline juju/model/__init__.py]: I could look and try to figure this out but I'll just be lazy and ask -- why the change from `self.applications` to `list(self.applications)`?

**Review by james-garner-canonical — COMMENTED** [review]
I haven't looked at the tests (though I noticed the addition of `freezegun`), but I've left some comments on the new implementation. My main concern is that the generator/coroutine business would be a lot easier (for me) to follow if we used a regular class/object instead (see comment on `def _loop`) -- it took me a while to get my head around. I do like the separation of the time-based checks. Lmk if you want me to look at anything else in particular

**Comment by dimaqq** [comment]
@james-garner-canonical I've refactored the loop to be an [async] filter.

It turns out that I've misunderstood the definition of "idle".

In short:

- ready units: depends on workload (and possibly agent, machine, etc.)
- idle units: depends on agent only

This means that I have to update the `idle.check` code.

Edit: done now.

File: juju/model/__init__.py (hunk 17 above)
Comment (dimaqq) [reply juju/model/__init__.py]: because of the type hint really, as self.applications is a dict, and keys are something special... anyway both are iterable, but type hints force some kind of cast.

File: juju/model/__init__.py (hunk 16 above)
Comment (dimaqq) [reply juju/model/__init__.py]: I've changed this to TypeError now.

File: juju/model/__init__.py (hunk 15 above)
Comment (dimaqq) [reply juju/model/__init__.py:3108]: yes, and I explicitly don't want to refactor the old idle loop, lest I break it and not notice.

File: juju/model/idle.py (hunk 14 above)
Comment (dimaqq) [reply juju/model/idle.py]: I'll have to rewrite this entirely.
https://matrix.to/#/!aEDlwVgqmZMLeZIcUP:ubuntu.com/$q0-_izqQi03-Eeqbpp259srutArMADqW2pVO_EDgIbY?via=ubuntu.com

File: juju/model/idle.py (hunk 13 above)
Comment (dimaqq) [reply juju/model/idle.py]: Maybe, I'll play with this a bit later.
There's a slight difference in logging, but that's not important.

File: juju/model/idle.py (hunk 14 above)
Comment (dimaqq) [reply juju/model/idle.py]: done

File: juju/model/__init__.py (hunk 16 above)
Comment (dimaqq) [reply juju/model/__init__.py]: done

File: juju/model/idle.py (hunk 13 above)
Comment (dimaqq) [reply juju/model/idle.py]: done

File: juju/model/idle.py (hunk 12 above)
Comment (dimaqq) [reply juju/model/idle.py]: https://matrix.to/#/!UPTldjWgQImSaCAjYf:ubuntu.com/$A20c0covFbezl2YplO2kqkGFvsp0n5-2Vnsrx1HiHao?via=ubuntu.com&via=matrix.org

File: juju/model/idle.py (hunk 11 above)
Comment (dimaqq) [reply juju/model/idle.py]: I've changed this to an async filter now.

Class code may be more straightforward, but I'll keep that as a backup plan.

File: juju/model/__init__.py (hunk 10 above)
Comment (dimaqq) [reply juju/model/__init__.py]: done, it's async filter now.

File: juju/model/idle.py (hunk 9 above)
Comment (dimaqq) [reply juju/model/idle.py]: good catch, done

File: juju/model.py (hunk 4 above)
Comment (dimaqq) [reply juju/model.py]: your reading of the old loop is pretty close, but I think I've found one caveat: https://matrix.to/#/!UPTldjWgQImSaCAjYf:ubuntu.com/$WRycclIa2h__O7A4vpJHuN4jBRygLeEQMYVS3gU3JFk?via=ubuntu.com&via=matrix.org

**Comment by dimaqq** [comment]
@james-garner-canonical idle implemented, most comments addressed, tests pass again. please review :)

File: juju/model/_idle.py [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+            ]
+            if len(ready_units) < wait_for_units:
+                logger.info(
+                    "Waiting for app %r units %s >= %s",
+                    app_name,
+                    len(status.ready_units),
+                    wait_for_units,
+                )
+                yield False
+                continue
```
Comment (dimaqq) [inline juju/model/_idle.py]: Ah, this is wrong, the intention is to continue the outer loop, not the inner!

Comment (dimaqq) [reply juju/model/_idle.py]: fixed and covered with tests.

File: juju/client/facade.py (hunk 8 above)
Comment (dimaqq) [reply juju/client/facade.py:498]: I've looked through the code here and in Connection.rpc() and I'm pretty sure that an "error" can't happen here.
I've added comments to that end.

File: juju/model/__init__.py [hunk 19]
```diff
@@ -3205,14 +3246,87 @@ def _raise_for_status(entities: dict[str, list[str]], status: Any):
...
+                    status=status,
+                )
+
+        async for done in _idle.loop(
+            status_on_demand(),
+            apps=apps,
+            wait_for_exact_units=wait_for_exact_units,
+            wait_for_units=wait_for_units,
+            idle_period=idle_period,
+        ):
```
Comment (james-garner-canonical) [inline juju/model/__init__.py:3320]: Suggestion: simplify this for noobs like me ... btw at first I thought this could be simplified without changing anything else, but after making it to `_idle.loop` I realised that my original simplification was totally incorrect, which suggests that the current version is a bit tricky to follow ... the below suggestion assumes a later suggestion to refactor `_idle.loop` to use a class, which I think makes things a lot easier to follow -- it certainly does for me
```suggestion
        timing_checker = _idle.TimingChecker()
        while True:
            full_status = await self.get_status()
            check_status = _idle.check(
                raw_status,
                apps=apps,
                raise_on_error=raise_on_error,
                raise_on_blocked=raise_on_blocked,
                status=status,
            )
            done = timing_checker.check(
                check_status,
                apps=apps,
                wait_for_exact_units=wait_for_exact_units,
                wait_for_units=wait_for_units,
                idle_period=idle_period,
            )
```

File: juju/model/_idle.py [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+
+    for unit_name, unit in units.items():
+        if unit.machine:
+            machine = full_status.machines[unit.machine]
+            assert isinstance(machine, MachineStatus)
+            assert machine.instance_status
+            if machine.instance_status.status == "error" and raise_on_error:
+                raise JujuMachineError(
+                    f"{unit_name!r} machine {unit.machine!r} has errored: {machine.instance_status.info!r}"
+                )
```
Comment (james-garner-canonical) [inline juju/model/_idle.py:137]: Suggestion: make it clear up front that we're only interested in unit.machine -- there's no elif/else coming
```suggestion
        if not unit.machine:
            continue
        machine = full_status.machines[unit.machine]
        assert isinstance(machine, MachineStatus)
        assert machine.instance_status
        if machine.instance_status.status == "error" and raise_on_error:
            raise JujuMachineError(
                f"{unit_name!r} machine {unit.machine!r} has errored: {machine.instance_status.info!r}"
            )
```

File: juju/model/_idle.py [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+    # Workload error (-"-)
+    # App error (any app from apps)
+    #
+    # Workload blocked (any unit of any app from apps)
+    # App blocked (any app from apps)
+    units: dict[str, UnitStatus] = {}
+
+    for app_name in apps:
+        units.update(_app_units(full_status, app_name))
+
```
Comment (james-garner-canonical) [inline juju/model/_idle.py:128]: Suggestion: I really like how the loops below all do one thing, much nicer to read and follow than the original wait_for_idle. I notice that the first few only do anything if `raise_on_error`, while the next couple only do something if `raise_on_blocked`. How about something like:

```python
    if raise_on_error:
        _check_for_errors(full_status, apps, units)
    if raise_on_blocked:
        _check_for_blocked(full_status, apps, units)
```

File: juju/model/_idle.py [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+            )
+
+    for app_name in apps:
+        app = full_status.applications[app_name]
+        assert isinstance(app, ApplicationStatus)
+        assert app.status
+        if app.status.status == "blocked" and raise_on_blocked:
+            raise JujuAppError(f"{app_name!r} is blocked: {app.status.info!r}")
+
+    rv = CheckStatus(set(), set(), set())
```
Comment (james-garner-canonical) [inline juju/model/_idle.py:174]: Minor suggestion: in the `CheckStatus` definition, use `= field(default_factory=set)`. Reasoning: I think `rv = CheckStatus()` will look more natural to readers (like me =))

File: juju/model/_idle.py [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+                logger.info(
+                    "Waiting for app %r units %s == %s",
+                    app_name,
+                    len(ready_units),
+                    wait_for_exact_units,
+                )
+                yield False
+                break
+        else:
+            yield True
```
Comment (james-garner-canonical) [inline juju/model/_idle.py:98]: Suggestion: I really think that a class to encapsulate the state, and a single synchronous method call call with returns, is so much easier to follow than the async generator, and it seems like it's pretty straightforwardly logically equivalent -- just a lot easier to follow for noobs like me

```suggestion
class TimingChecker:
    """The outer, time-dependent logic of a wait_for_idle loop."""
    def __init__(self):
        self.idle_since: dict[str, float] = {}
    
    def check(
        status: CheckStatus,
        *
        apps: AbstractSet[str],
        wait_for_exact_units: int | None = None,
        wait_for_units: int,
        idle_period: float,
    ) -> bool:
        logger.info("wait_for_idle iteration %s", status)
        now = time.monotonic()

        if not status:
            return False

        expected_idle_since = now - idle_period

        # FIXME there's some confusion about what a "busy" unit is
        # are we ready when over the last idle_period, every time sampled:
        # a. >=N units were ready (possibly different each time), or
        # b. >=N units were ready each time
        for name in status.units:
            if name in status.idle_units:
                self.idle_since[name] = min(now, idle_since.get(name, float("inf")))
            else:
                self.idle_since[name] = float("inf")

        if busy := {n for n, t in self.idle_since.items() if t > expected_idle_since}:
            logger.info("Waiting for units to be idle enough: %s", busy)
            return False

        for app_name in apps:
            ready_units = [
                n for n in status.ready_units if n.startswith(f"{app_name}/")
            ]
            if len(ready_units) < wait_for_units:
                logger.info(
                    "Waiting for app %r units %s >= %s",
                    app_name,
                    len(status.ready_units),
                    wait_for_units,
                )
                return False
            if (
                wait_for_exact_units is not None
                and len(ready_units) != wait_for_exact_units
            ):
                logger.info(
                    "Waiting for app %r units %s == %s",
                    app_name,
                    len(ready_units),
                    wait_for_exact_units,
                )
                return False
        return True
```

**Review by james-garner-canonical — CHANGES_REQUESTED** [review]
Exciting updates, glad tests are passing now.

I'm finding it hard to follow the logical changes between versions on this PR, I think primarily because the generator adds a fair bit of mental overhead for me -- it's kind of nicer now (though the new outer async loop construct took me some time to digest), but I still have to put in more mental work to follow the various `yield ... break`s and `yield ... continue`s.

For now I've made some requests for changes to make it easier for me to follow what's going on. Sorry if that seems like wasted effort -- it did take me a while, but I kind of had to do it anyway to follow the code.

Let me know if you want me to look closer at specific logical changes (looks like there are changes to the `idle_since` logic?). I still haven't looked in detail at your tests either, would you like further review there?

Also, what's the latest with running charmers' integration tests on the latest version? Are there still failures with the new wait_for_idle that pass with the old one?

**Comment by dimaqq** [comment]
I've pushed a separate branch with this refactor only: https://github.com/juju/python-libjuju/pull/1245 PTAL

**Comment by dimaqq** [comment]
closing in favour of #1245
